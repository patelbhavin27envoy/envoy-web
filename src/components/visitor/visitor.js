import React, {Component} from 'react';
import entryService from '../../services/entryService';
import './visitor.scss';

class Visitor extends Component {

    constructor(props) {
        super(props)
        this.state = {
            signOutText : 'Sign Out',
            visitor : this.props.visitor,
            loading : false,
            error: false
        };
        this.signOut = this.signOut.bind(this);
        this.callService = this.callService.bind(this);
    }

    signOut() {

        // set state for the button
        this.setState({
            loading: true,
            error: false
        });

        setTimeout(this.callService(), 100000);
    }

    callService() {

        entryService()
            .signOutVisitor(this.state.visitor._id)
            .then((data) => {

                if(data && data.status === true) {
                    return this.setState({
                        loading: false,
                        visitor: data.visitor,
                        error: false
                    });
                } else {
                    this.setState({
                        error: true,
                        'loading' : false
                    })
                }
                // This is for Redux
                // this.props.venuesReceived(data);

            })
            .catch((error) => {

                return this.setState({
                    error: true,
                    'loading' : false
                })
            });
    }

    render() {

        let { visitor : { name, notes, signOutTime, _id, signedOut, signedIn}, error} = this.state;

        let signOutColumnElement = <button className="btn btn-smaller btn--outline" onClick={this.signOut}>
                                        Sign out
                                    </button>;
        if(signedOut === true && signedIn === true) {
            signOutColumnElement = signOutTime;
        }

        if(this.state.loading === true) {
            signOutColumnElement = "loading ...";
        }

        return (
                    <tr>
                        <td className="p-2 border-t border-grey-light font-mono text-xs one">{name} </td>
                        <td className="p-2 border-t border-grey-light font-mono text-xs two">{notes}</td>
                        <td className="p-1 border-t border-grey-light font-mono text-xs three">
                            {signOutColumnElement}

                            {this.state.error && (
                                    <div className='error red'>
                                            <div>Error Updating!!</div>
                                            <div >Try later!!</div>
                                    </div>
                                )
                            }
                        </td>
                    </tr>
        )
    }
}

export default Visitor;
