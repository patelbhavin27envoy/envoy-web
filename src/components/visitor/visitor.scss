table {
  thead {
      th {
        text-align: left;
      }

      td.one,
      td.two,
      td.three {
        width: 33.3%;
      }
  }


  tbody {
    th {
      text-align: left;
    }
    td.one,
    td.two,
    td.three {
      width: 33.3%;
    }
  }

}
