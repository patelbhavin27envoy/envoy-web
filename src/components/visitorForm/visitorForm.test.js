import React from 'react';
import renderer from 'react-test-renderer';

import { shallow, mount } from "enzyme";
import { MemoryRouter } from 'react-router-dom';

import configureMockStore from "redux-mock-store";

import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import { VisitorForm } from '../VisitorForm'
import Search from "../Search/Search";
import {Visitor} from "../visitor";

let visitors = [{
        firstName : 'John',
        lastName : 'Doe',
        notes : 'test first',
        signedOutDate : new Date('02/09/2019').toDateString()
    },
    {
        firstName : 'Esteban',
        lastName : 'Arango',
        notes : 'Frisbee and Vegan food',
        signedOutDate : new Date('04/24/2019').toDateString()
    },
    {
        firstName : 'Ryan',
        lastName : 'Labouve',
        notes : 'Everything about Oklahoma',
        signedOutDate : null
    }];

// For Redux
Enzyme.configure({ adapter: new Adapter() });
const mockStore = configureMockStore();
const store = mockStore({
    venues : visitors
});


describe("VisitorForm Component :: ", () => {

    it("should render", () => {

        // Act
        let component = renderer.create(
            <MemoryRouter>
            <VisitorForm />
            </MemoryRouter>
    ).toJSON();
        // snapshot
        expect(component).toMatchSnapshot();


        // Act
        component = mount(
            <MemoryRouter>
            <VisitorForm />
            </MemoryRouter>
    );

        // Assert
        expect((component).prop('state')).not.toEqual(null);
        expect(component.find('form')).not.toEqual(null);
        expect(component.find('.py3.block')).not.toEqual(null);
        expect(component.find('.form-name').text()).toContain('Welcome to Envoy!!');
        expect(component.find('.new-visitor').text()).toContain('Add Details For Your Visit.');
        expect(component.find('.field').text()).toContain('First Name');
        expect(component.find('.field').text()).toContain('Last Name');
        expect(component.find('.field').text()).toContain('Note');

        component.find('.firstName').simulate('change', {target: {name: 'firstName', value: 'John'}});
        component.find('.lastName').simulate('change', {target: {name: 'lastName', value: 'Doe'}});
        component.find('.note').simulate('change', {target: {name: 'note', value: 'note'}});

    });

    it("should check goBack link", () => {

         // Act
        const component = mount(
            <MemoryRouter initialEntries={[ '/visitors']}>
                <VisitorForm />
            </MemoryRouter>
        );

        // Assert
        expect((component).prop('state')).not.toEqual(null);
        component.find('.go-back').simulate('click');

        // const instance = component.instance();
        // expect((component).prop('goBack')).not.toEqual(null);
        // jest.spyOn(instance, (instance).prop('goBack'));
        // instance.componentDidMount();
        // expect(instance.func).toHaveBeenCalled();
        //
        // expect(methodNameFake).toHaveBeenCalledTimes(1);
    });



    it("should check reset and save :: Failure ", () => {

        // define the response
        const mockSuccessResponse = {status : true, visitor : [visitors[2]]};

        // let's mock the service call
        const endPoint = 'http://localhost:8080/api/visitors';
        const mockJsonPromise = Promise.resolve(mockSuccessResponse);
        const mockFetchPromise = Promise.resolve({
            json: () => mockJsonPromise,
        });
        jest.spyOn(global, 'fetch').mockImplementation(() => { return mockFetchPromise }); // 4

        // Act
        const component = mount(
            <MemoryRouter >
                <VisitorForm />
            </MemoryRouter>
         );


        component.find('.firstName').simulate('change', {target: {name: 'firstName', value: 'John'}});
        component.find('.lastName').simulate('change', {target: {name: 'lastName', value: 'Doe'}});
        component.find('.note').simulate('change', {target: {name: 'note', value: 'note'}});

        component.find('.reset').simulate('click');
        component.find('.save').simulate('click');

        // expect(component.state('error')).toBe(false);
        expect(component).toMatchSnapshot();

    });


    it("should check SAVE :: success ", () => {

        // define the response
        const mockSuccessResponse = {status : true, visitor : [visitors[2]]};

        // let's mock the service call
        const endPoint = 'http://localhost:8080/api/visitors';
        const mockJsonPromise = Promise.resolve(mockSuccessResponse);
        const mockFetchPromise = Promise.resolve({
            json: () => mockJsonPromise,
        });
        jest.spyOn(global, 'fetch').mockImplementation(() => { return mockFetchPromise }); // 4

        // Act
        const component = mount(
            <MemoryRouter >
            <VisitorForm />
            </MemoryRouter>
        );

        component.find('.firstName').simulate('change', {target: {name: 'firstName', value: 'John'}});
        component.find('.lastName').simulate('change', {target: {name: 'lastName', value: 'Doe'}});
        component.find('.note').simulate('change', {target: {name: 'note', value: 'note'}});

        component.find('.save').simulate('click');

        // expect(component.state('error')).toBe(false);
        expect(component).toMatchSnapshot();

    });


    it("should make API call :: Failure-2 ", () => {

        // define the response
        // const mockSuccessResponse = {};

        // let's mock the service call
        const endPoint = 'http://localhost:8080/api/visitors/';
        const mockJsonPromise = Promise.resolve({data : { statue: false} });
        const mockFetchPromise = Promise.resolve({
            json: () =>  mockJsonPromise
        });
        jest.spyOn(global, 'fetch').mockImplementation(() => { return mockFetchPromise }); // 4


        // Act
        const component = mount(
            <MemoryRouter >
                <VisitorForm />
            </MemoryRouter>
        );

        component.find('.firstName').simulate('change', {target: {name: 'firstName', value: 'John'}});
        component.find('.lastName').simulate('change', {target: {name: 'lastName', value: 'Doe'}});
        component.find('.note').simulate('change', {target: {name: 'note', value: 'note'}});

        component.find('.save').simulate('click');

        // expect(component.state('error')).toBe(false);
        expect(component).toMatchSnapshot();
    });

});

