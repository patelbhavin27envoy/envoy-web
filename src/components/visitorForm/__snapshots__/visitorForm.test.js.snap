// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VisitorForm Component ::  should check SAVE :: success  1`] = `ReactWrapper {}`;

exports[`VisitorForm Component ::  should check reset and save :: Failure  1`] = `ReactWrapper {}`;

exports[`VisitorForm Component ::  should make API call :: Failure-2  1`] = `ReactWrapper {}`;

exports[`VisitorForm Component ::  should render 1`] = `
Array [
  <div
    className="header"
  >
    <img
      alt="Envoy Logo"
      className="py3 block logo"
      src="https://dashboard.envoy.com/assets/images/logo-small-red-ba0cf4a025dd5296cf6e002e28ad38be.svg"
      width="31"
    />
    <div
      className="form-name ml-2"
    >
      Welcome to Envoy!!
    </div>
  </div>,
  <div
    className="new-visitor"
  >
    <h3
      className="details"
    >
      Add Details For Your Visit.
    </h3>
    <form
      className="field"
    >
      <div
        className="element-container"
      >
        <span
          className="tag"
        >
          First Name 
        </span>
        <input
          className="firstName p-2 text-sm border  w-full"
          name="firstName"
          onChange={[Function]}
          placeholder="First Name"
          type="text"
          value=""
        />
      </div>
      <div
        className="element-container"
      >
        <span
          className="tag"
        >
          Last Name 
        </span>
        <input
          className="lastName p-2 text-sm border  w-full"
          name="lastName"
          onChange={[Function]}
          placeholder="Last Name"
          type="text"
          value=""
        />
      </div>
      <div
        className="element-container"
      >
        <span
          className="tag"
        >
          Note
        </span>
        <textarea
          className="note p-2 text-sm border w-full"
          cols="40"
          name="lastName"
          onChange={[Function]}
          placeholder="What brings you here today?"
          rows="4"
          value=""
        />
      </div>
      <div
        className="btn-container"
      >
        <a
          className="float-left go-back redLighter"
          href="#"
          onClick={[Function]}
        >
           Back 
        </a>
        <button
          className="btn  btn--brand ml-2 float-right save"
          onClick={[Function]}
        >
          <i
            className="fas fa-user"
          />
            Save
        </button>
        <button
          className="btn  btn--brand ml-2 float-right reset"
          onClick={[Function]}
        >
          reset
        </button>
      </div>
    </form>
  </div>,
]
`;
